scripts/src/sampleAndSaveMarkdownPages.ts

```diff
@@ -0,0 +1,89 @@
+import { MongoClient, ObjectId } from "mongodb";
+import { assertEnvVars } from "chat-core";
+import { promises as fs } from "fs";
+import path from "path";
+
+import "dotenv/config";
+
+const SAMPLE_SIZE = 100;
+const OUTPUT_DIRECTORY = "./sampled-pages";
+
+const { MONGODB_DATABASE_NAME, MONGODB_CONNECTION_URI } = assertEnvVars({
+  MONGODB_DATABASE_NAME: "",
+  MONGODB_CONNECTION_URI: "",
+});
+
+type Page = {
+  _id: ObjectId;
+  sourceName: string;
+  url: string;
+  action: string;
+  body: string;
+  updated: Date;
+  title?: string;
+  tags?: string | string[];
+  metadata?: object;
+};
+
+async function prepareOutputDirectory() {
+  const res = await fs.mkdir(OUTPUT_DIRECTORY, { recursive: true });
+  return res;
+}
+
+async function savePageMarkdown(page: Page) {
+  const res = await fs.writeFile(
+    path.join(
+      OUTPUT_DIRECTORY,
+      `${page.sourceName}_${page.title ?? page._id}.md`
+    ),
+    page.body
+  );
+  return res;
+}
+
+function partition<T, Pass extends T = T, Fail extends T = T>(
+  arr: T[],
+  test: (elem: T) => boolean
+) {
+  type PassFail = [Pass[], Fail[]];
+  const passes: Pass[] = [];
+  const fails: Fail[] = [];
```

Comment:
I think you create possible pain for yourself declaring these locally and having them shadowed within the reducer. I would just put `baseAccumulator = [[], []];`